### PR TITLE
Fix DLL loading

### PR DESCRIPTION
**Pull request by TheCurle**
Export shared_ptr<GameModule> from the DLL, and store the dylib in a wider scope.

Additionally, all modules are shared_ptr to prevent the Game Module from going out of scope.

Additionally, the Module needs to be pure virtual.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -18,7 +18,7 @@ FetchContent_Declare(
         GIT_TAG        v2.13.9 # or a later release
 )
 
-FetchContent_MakeAvailable(Catch2)
```
Comment (dpeter99) [inline CMakeLists.txt]: why did you remove the Catch2 lib?

File: CMakeLists.txt [hunk 2]
```diff
@@ -75,7 +75,7 @@ project(umbra)
 add_subdirectory(projs/shadow/shadow-utility)
 
 # Asset management
-add_subdirectory(projs/shadow/shadow-file-format)
```
Comment (dpeter99) [inline CMakeLists.txt]: Why did you remove the ``file-format`` proj?

File: projs/shadow/shadow-engine/inc/core/ModuleManager.h [hunk 3]
```diff
@@ -9,7 +9,7 @@ namespace ShadowEngine {
 
     struct ModuleRef{
     public:
-        Module* module;
+        std::shared_ptr<Module> module;
```
Comment (dpeter99) [inline projs/shadow/shadow-engine/inc/core/ModuleManager.h]: The module can't be a shared pointer cause for hot reloading we need to forcefully unload it.

File: projs/test-game/inc/GameModule.h [hunk 4]
```diff
@@ -1,9 +1,4 @@
-//
-// Created by dpete on 02/09/2022.
-//
-
-#ifndef UMBRA_GAMEMODULE_H
-#define UMBRA_GAMEMODULE_H
+#pragma once
```
Comment (dpeter99) [inline projs/test-game/inc/GameModule.h:1]: Isn't ``pragma once`` msvc only?

**Review by dpeter99 — CHANGES_REQUESTED** [review]
This is nice!
After trying to compile it I got the following error:
```
error C2526: 'shadow_main': C linkage function cannot return C++ class 'std::shared_ptr<ShadowEngine::Module>'
```

Does this currently solve calling the instance get on ``ShadowApplication`` ?

File: CMakeLists.txt (hunk 1 above)
Comment (TheCurle) [reply CMakeLists.txt]: Catch2 and the file format project were temporarily removed because they require Python and I couldn't be bothered installing it.

File: projs/shadow/shadow-engine/inc/core/ModuleManager.h (hunk 3 above)
Comment (TheCurle) [reply projs/shadow/shadow-engine/inc/core/ModuleManager.h]: We can forcefully delete the shared pointer. However, we'd need to tell the dll itself to do it.

File: projs/test-game/inc/GameModule.h (hunk 4 above)
Comment (TheCurle) [reply projs/test-game/inc/GameModule.h:1]: Given that I'm using mingw to compile, no.

**Comment by TheCurle** [comment]
I'm not sure why you're getting that error on the entry.cpp file, it's completely valid by spec for a C linkage function to return a C++ class...

And yeah. By making the dll return just the instance of itself, rather than reaching out into the Shadow Application to insert itself, the dependency loop (engine imports game imports engine imports game imports...) is solved and the game can be linked.
